osu.Game/Updater/AssociationManager.cs

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
+using Microsoft.Win32;
+using osu.Framework.Allocation;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Logging;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+
+namespace osu.Game.Updater
+{
+    [SupportedOSPlatform("windows")]
+    public partial class AssociationManager : CompositeDrawable
+    {
+        // Needed so that Explorer windows get refreshed after the registry is updated
+        [DllImport("Shell32.dll")]
+        private static extern int SHChangeNotify(int eventId, int flags, IntPtr item1, IntPtr item2);
+
+        private static readonly string[] associated_extensions =
+        {
+            ".osz",
+            ".osr",
+            ".osk",
+            ".osu",
+            ".osr",
+        };
+
+        [Resolved]
+        protected INotificationOverlay Notifications { get; private set; } = null!;
+
+        // TODO: Somehow make this run on initial setup without prompting the user
+        /// <summary>
+        /// This checks the integrity of the file associations. If they look out of place, it will prompt the user via notifications to fix them.
+        /// </summary>
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Logger.Log("Checking file associations");
+
+            bool fileAssociationStatus = EnsureAssociationsSet();
+
+            if (!fileAssociationStatus)
+            {
+                // Files are not associated, prompt the user to fix them
+                Notifications.Post(new SimpleNotification
```

Comment:
> I can't think of a situation where the user would want osu! to be installed and not handle these extensions.

A scenario, or really the only scenario, would be a person who primarily uses stable but has lazer installed alongside stable to simply try out lazer.

My original idea was to have it notify the user on first time setup, and allow them to do it there. If the user declines, then they can simply create the associations themselves within the maintenance section of the settings menu at their own convenience. However, I proposed the idea of having lazer check the integrity of the associations on startup to make sure they look fine, which you were on board with, so I went that route instead of having it notify every time on startup if the associations look out of place.

> Also means that this component could potentially be static and non-drawable as a result.

That would be a huge plus... But I can't really decide at the moment, I'll need to think it over and see what other approaches I could take when it comes to this. One side of me feels like allowing the user to have control over this would be a good idea, but another half agrees that automating this for the user could also be a good idea. I wonder if it would be possible to poll this or something?

Reply:
> A scenario, or really the only scenario, would be a person who primarily uses stable but has lazer installed alongside stable to simply try out lazer.

If stable always takes priority, then the only thing we'd want is a way to uninstall stable's handlers. I don't think installing the lazer ones needs any prompting.

Reply:
That's certainly an idea. But to have lazer remove stable's would require elevation. It isn't that big of an issue though, I can definitely see this being a method.

I do have an idea for a method without any prompting:
Slap it into the settings menu, and have a dropdown where you can select which installation to point associations to (stable, lazer). When you select lazer, it uninstalls stable's keys (if needed) and implements lazer's own keys. When you select stable, it reimplements stable's keys.

That allows people to choose where associations should point to, and allows us to set lazer as the default, without prompting the user as you wish.